### PR TITLE
Add API to mark input tensors static for cudagraphs

### PR DESCRIPTION
**Pull request by mlazos**
Adds API to mark tensor as a static input -
To make this trigger recompiles properly, I'll need to update tensor match checks to also check for this new attribute

Additional concern is memory - the tensors will be kept alive, but this is the current behavior for nn modules and parameters.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

File: torch/_dynamo/output_graph.py [hunk 1]
```diff
@@ -607,9 +608,13 @@ def register_attr_or_module(
             if not is_constant_source(source):
                 options["guards"].add(source.make_guard(GuardBuilder.TENSOR_MATCH))
 
+            if get_static_input_type(target) == "guarded":
+                options["guards"].add(source.make_guard(GuardBuilder.ID_MATCH))
```
Comment (eellison) [inline torch/_dynamo/output_graph.py:612]: Can we do DATA_PTR match instead ?

File: torch/_dynamo/decorators.py [hunk 2]
```diff
@@ -236,6 +236,20 @@ def mark_static(t, index=None):
             mark_static(t, i)
 
 
+@forbid_in_graph
+def mark_static_input(t, guard=True):
```
Comment (eellison) [inline torch/_dynamo/decorators.py:240]: We could also do mark_static_address ?

**Review by eellison — COMMENTED** [review]
Looks good, but right now we have a mismatch between the api, which says its guarding on data_ptr, versus the guard, which is on ID>

The concern I have is if you do:

```
for _ in range(3):
    torch._inductor.config.cudagraphs = True

    @torch.compile()
    def foo():
        return torch.rand([4])
        
    out = foo()
    torch._dynamo.mark_static_input(out)

    @torch.compile()
    def foo2(x):
        return x + x
```

a) this is actually unnecessary, bc cudagraphs will itself guard that out is always static address.

but more importantly 

b) This is actually going to cause recompilations. Cudagraphs will use the a different tensor id from run 1-> 2 -> 3. The first run is a warm up. The second run we are recording the graphs, so we need the tensors to die so their memory is reclaimed. It's only when we hit path 3, that we actually cache the output tensors and the id becomes consistent from run to run.

**Comment by eellison** [comment]
cc @Chillee

File: torch/_dynamo/output_graph.py (hunk 1 above)
Comment (mlazos) [reply torch/_dynamo/output_graph.py:612]: Yeah sure, aren't they equivalent tho?

File: torch/_dynamo/decorators.py (hunk 2 above)
Comment (mlazos) [reply torch/_dynamo/decorators.py:240]: yeah I thought this name for a while, address is better, ty!

File: torch/_dynamo/decorators.py [hunk 3]
```diff
@@ -236,6 +236,24 @@ def mark_static(t, index=None):
             mark_static(t, i)
 
 
+@forbid_in_graph
+def mark_static_address(t, guard=True):
+    """
+    Marks an input tensor whose data_ptr will not change across multiple calls
+    to a dynamo-compiled function. This indicates to cudagraphs that an extra allocation
+    is not needed for this input. The data_ptr will be guarded if guard=True. Note:
+    Until https://github.com/pytorch/pytorch/issues/104095 is resolved, tensors marked
```
Comment (eellison) [inline torch/_dynamo/decorators.py:245]: Even if https://github.com/pytorch/pytorch/issues/104095, is partially fixed with the ID_Match thing weakref thing, the tensors would be kept alive if this invoked on a free function. 

And we dont have any particular proposal of how clean-up should occur with free functions.

Maybe better to just say these are kept alive ?

File: torch/_dynamo/decorators.py [hunk 4]
```diff
@@ -236,6 +236,24 @@ def mark_static(t, index=None):
...
+    in this way will be kept alive until `torch._dynamo.reset()` is called.
+    """
+    if not isinstance(t, torch.Tensor):
+        raise ValueError("mark_static_address expects a tensor")
```
Comment (eellison) [inline torch/_dynamo/decorators.py:249]: nit: `TypeError`  & `expects a tensor but got {type(t)}`

**Review by eellison — APPROVED** [review]
looks great !

File: torch/_dynamo/decorators.py (hunk 3 above)
Comment (mlazos) [reply torch/_dynamo/decorators.py:245]: Yeah I can do that, I'll likely have to do the finalizer approach for the optimizer

**Comment by ezyang** [comment]
Subbing myself

**Pull request closed without merging**